% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/methods.R
\name{create_scCrossTalk}
\alias{create_scCrossTalk}
\title{Create scCrossTalk object}
\usage{
create_scCrossTalk(sc_data, sc_celltype, species, if_normalize = TRUE)
}
\arguments{
\item{sc_data}{A data.frame or matrix or dgCMatrix containing raw counts of single-cell RNA-seq data. see \code{\link{demo_sc_data}}}

\item{sc_celltype}{A character containing the cell type of the single-cell RNA-seq data}

\item{species}{A character meaning species of the single-cell transcriptomics data.\code{'Human'} or \code{'Mouse'}.}

\item{if_normalize}{Normalize sc_data with Seurat LogNormalize. Set it \code{FLASE} when sc_data has been normalized.}
}
\value{
scCrossTalk object
}
\description{
create scCrossTalk object using single-cell transcriptomics data
}
